crates/scsi/src/main.rs

```diff
@@ -0,0 +1,400 @@
+            // rebooting. So let's just log an error and do nothing.
+            error!("Error writing response to guest memory: {:?}", e);
+        }
+    }
+
+    fn add_target(&mut self, target: Box<Target>) {
+        self.targets.push(target);
+    }
+}
+
+impl VhostUserBackendMut<VringRwLock> for VhostUserScsiBackend {
+    fn num_queues(&self) -> usize {
+        let num_request_queues = 1;
+        2 + num_request_queues
+    }
+
+    fn max_queue_size(&self) -> usize {
+        128 // qemu assumes this by default
+    }
+
+    fn features(&self) -> u64 {
+        // TODO: Any other ones worth implementing? EVENT_IDX and INDIRECT_DESC
+        // are supported by virtiofsd
+        1 << VIRTIO_F_VERSION_1 | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() | 1 << 2
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ
+    }
+
+    fn set_event_idx(&mut self, enabled: bool) {
+        // Should always be true until we support EVENT_IDX in features.
+        assert!(!enabled);
+    }
+
+    fn update_memory(
+        &mut self,
+        atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    ) -> std::result::Result<(), std::io::Error> {
+        info!("Memory updated - guest probably booting");
+        self.mem = Some(atomic_mem);
+        Ok(())
+    }
+
+    fn handle_event(
+        &mut self,
+        device_event: u16,
+        evset: EventSet,
+        vrings: &[VringRwLock],
+        thread_id: usize,
+    ) -> io::Result<bool> {
+        assert!(evset == EventSet::IN);
+        assert!(vrings.len() == 3);
+        assert!((device_event as usize) < vrings.len());
+        assert!(thread_id == 0);
+
+        // unwrap: only fails if the lock is poisoned, in which case we already panicked
+        // somewhere else
+        let mut vring = vrings[device_event as usize].get_mut();
+        let queue = vring.get_queue_mut();
+
+        let chains: Vec<_> = queue
+            .iter(self.mem.as_ref().unwrap().memory())
+            .map_err(|e| io::Error::new(ErrorKind::Other, e))?
+            .collect();
+
+        for dc in chains {
+            let mut writer = DescriptorChainWriter::new(dc.clone());
+            let mut reader = DescriptorChainReader::new(dc.clone());
+
+            #[allow(clippy::single_match_else)]
+            match device_event {
+                2 => self.handle_request_queue(&mut reader, &mut writer),
+                _ => {
+                    error!("Ignoring descriptor on queue {}", device_event);
+                    continue;
+                }
+            }
+
+            vring
+                .add_used(dc.head_index(), writer.max_written())
+                .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
+        }
+
+        vring
+            .signal_used_queue()
+            .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
+
+        Ok(false) // TODO: what's this bool? no idea. virtiofd-rs returns false
```

Comment:
review

Reply:
clarification happens at: https://github.com/rust-vmm/vhost/issues/144. I guess this is fine and I can drop the comment.